File: langs/javascript/src/typed.ts
/**
 * TypeScript-first typed configuration API for DOTLYTE.
 *
 * Provides `createTypedConfig()` with full TypeScript generic type inference,
 * server/client boundary enforcement, and Zod/Valibot schema support.
 *
 * @example
 * ```ts
 * import { createTypedConfig } from 'dotlyte'
 *
 * const env = createTypedConfig({
 *   DATABASE_URL: { type: 'string', format: 'url', required: true },
 *   PORT: { type: 'integer', default: 3000 },
 *   DEBUG: { type: 'boolean', default: false },
 *   LOG_LEVEL: { type: 'string', enum: ['debug', 'info', 'warn', 'error'] as const, default: 'info' },
 *   REDIS_URL: { type: 'string', format: 'url', required: false },
 * })
 *
 * env.DATABASE_URL  // string (never undefined — required)
 * env.PORT          // number (never undefined — has default)
 * env.DEBUG         // boolean
 * env.LOG_LEVEL     // 'debug' | 'info' | 'warn' | 'error'
 * env.REDIS_URL     // string | undefined
 * ```
 *
 * @module dotlyte/typed
 */

import { load } from "./loader.js";
import { DotlyteError, ValidationError } from "./errors.js";
import type { LoadOptions } from "./loader.js";
import { createBoundaryProxy } from "./boundaries.js";

/* ──────── Schema Descriptor Types ──────── */

/** Primitive type names supported in DotlyteSchema descriptors. */
export type TypeName = "string" | "integer" | "number" | "boolean" | "url";

/** A single field descriptor in the typed config schema. */
export interface FieldDescriptor<
  T extends TypeName = TypeName,
  E extends readonly unknown[] = readonly unknown[],
  D = undefined,
  R extends boolean = true,
> {
  /** The expected type. */
  type: T;
  /** Format constraint (e.g., 'url', 'email'). */
  format?: string;
  /** Whether this field is required. Defaults to `true`. */
  required?: R;
  /** Default value. If provided, the field is never undefined. */
  default?: D;
  /** Allowed values. Narrows the inferred type to a union. */
  enum?: E;
  /** Minimum value (numbers) or length (strings). */
  min?: number;
  /** Maximum value (numbers) or length (strings). */
  max?: number;
  /** Mark as sensitive — value will be redacted in logs. */
  sensitive?: boolean;
  /** Human-readable description. */
  doc?: string;
}

/* ──────── Type Inference Engine ──────── */

/** Map type names to their TypeScript types. */
type TypeNameToTS<T extends TypeName> = T extends "string" | "url"
  ? string
  : T extends "integer" | "number"
    ? number
    : T extends "boolean"
      ? boolean
      : unknown;

/** Infer the TypeScript type from a single field descriptor. */
type InferField<F> =
  F extends { enum: readonly (infer E)[] }
    ? F extends { default: infer _D }
      ? E
      : F extends { required: false }
        ? E | undefined
        : E
    : F extends { type: infer T extends TypeName }
      ? F extends { default: infer _D }
        ? TypeNameToTS<T>
        : F extends { required: false }
          ? TypeNameToTS<T> | undefined
          : TypeNameToTS<T>
      : unknown;

/** Check if a value is a Zod-like schema (has _def property). */
type IsZodSchema<T> = T extends { _def: unknown; parse: (...args: unknown[]) => unknown } ? true : false;

/** Check if a value is a Valibot-like schema. */
type IsValibotSchema<T> = T extends { type: string; _parse: unknown } ? true : false;

/** Infer type from a Zod schema. */
type InferZodField<T> = T extends { _output: infer O } ? O : unknown;

/** Infer type from a Valibot schema. We rely on the pipe mechanism. */
type InferValibotField<T> = T extends { type: "optional"; wrapped: infer W }
  ? InferValibotField<W> | undefined
  : T extends { type: "string" }
    ? string
    : T extends { type: "number" }
      ? number
      : T extends { type: "boolean" }
        ? boolean
        : unknown;

/**
 * Infer the type for a single field, supporting DotlyteSchema, Zod, and Valibot.
 */
type InferAnyField<F> =
  IsZodSchema<F> extends true
    ? InferZodField<F>
    : IsValibotSchema<F> extends true
      ? InferValibotField<F>
      : InferField<F>;

/** Infer the full config type from a flat schema object. */
export type InferConfig<S extends Record<string, unknown>> = {
  readonly [K in keyof S]: InferAnyField<S[K]>;
};

/* ──────── Sectioned Schema Types ──────── */

/** Schema with server/client/shared sections. */
export interface SectionedSchema<
  S extends Record<string, unknown> = Record<string, unknown>,
  C extends Record<string, unknown> = Record<string, unknown>,
  H extends Record<string, unknown> = Record<string, unknown>,
> {
  server: S;
  client: C;
  shared?: H;
  /** Prefix required for client variables (default: 'NEXT_PUBLIC_'). */
  clientPrefix?: string;
}

/** Detect whether a schema is sectioned. */
type IsSectionedSchema<S> = S extends { server: Record<string, unknown>; client: Record<string, unknown> }
  ? true
  : false;

/** Infer the combined config type from a sectioned schema. */
type InferSectionedConfig<S> = S extends SectionedSchema<infer Srv, infer Cli, infer Sh>
  ? InferConfig<Srv> & InferConfig<Cli> & InferConfig<Sh>
  : never;

/** The full inferred type: flat or sectioned depending on input. */
export type InferTypedConfig<S> =
  IsSectionedSchema<S> extends true ? InferSectionedConfig<S> : InferConfig<S & Record<string, unknown>>;

/* ──────── Options ──────── */

/** Options for createTypedConfig. */
export interface TypedConfigOptions extends Omit<LoadOptions, "schema"> {
  /**
   * Skip validation (useful in tests).
   * @default false
   */
  skipValidation?: boolean;
  /**
   * Callback invoked when a sensitive key is accessed.
   * Useful for audit logging (SOC2/HIPAA compliance).
   */
  onSecretAccess?: (key: string, context: string) => void;
}

/* ──────── Runtime Validators ──────── */

/** Validate and coerce a value against a DotlyteSchema field descriptor. */
function validateField(key: string, raw: unknown, descriptor: FieldDescriptor<TypeName, readonly unknown[], unknown, boolean>): unknown {
  const isRequired = descriptor.required !== false;
  const hasDefault = descriptor.default !== undefined;

  // Apply default
  let value = raw;
  if ((value === undefined || value === null || value === "") && hasDefault) {
    value = descriptor.default;
  }

  // Check required
  if ((value === undefined || value === null || value === "") && isRequired) {
    throw new DotlyteError(
      `Missing required environment variable '${key}'.` +
        (descriptor.doc ? ` (${descriptor.doc})` : "") +
        ` Set it in your .env file, config file, or as an environment variable.`,
      key,
      "MISSING_REQUIRED_KEY",
    );
  }

  // Optional and absent
  if (value === undefined || value === null || value === "") {
    return undefined;
  }

  // Coerce to target type
  const strValue = String(value);

  switch (descriptor.type) {
    case "boolean": {
      if (typeof value === "boolean") break;
      const lower = strValue.toLowerCase();
      if (["true", "yes", "1", "on"].includes(lower)) {
        value = true;
      } else if (["false", "no", "0", "off"].includes(lower)) {
        value = false;
      } else {
        throw new DotlyteError(
          `Environment variable '${key}' expected boolean, got '${strValue}'.`,
          key,
          "VALIDATION_ERROR",
        );
      }
      break;
    }
    case "integer":
    case "number": {
      if (typeof value === "number") break;
      const num = Number(strValue);
      if (isNaN(num)) {
        throw new DotlyteError(
          `Environment variable '${key}' expected number, got '${strValue}'.`,
          key,
          "VALIDATION_ERROR",
        );
      }
      if (descriptor.type === "integer" && !Number.isInteger(num)) {
        throw new DotlyteError(
          `Environment variable '${key}' expected integer, got '${strValue}'.`,
          key,
          "VALIDATION_ERROR",
        );
      }
      value = num;
      break;
    }
    case "string":
    case "url":
      value = strValue;
      break;
  }

  // Format validation
  if (descriptor.format === "url" || descriptor.type === "url") {
    try {
      new URL(String(value));
    } catch {
      throw new DotlyteError(
        `Environment variable '${key}' is not a valid URL: '${String(value)}'.`,
        key,
        "VALIDATION_ERROR",
      );
    }
  }

  // Enum validation
  if (descriptor.enum && !descriptor.enum.includes(value)) {
    throw new DotlyteError(
      `Environment variable '${key}' must be one of [${descriptor.enum.join(", ")}], got '${String(value)}'.`,
      key,
      "VALIDATION_ERROR",
    );
  }

  // Min/max validation
  if (typeof value === "number") {
    if (descriptor.min !== undefined && value < descriptor.min) {
      throw new DotlyteError(
        `Environment variable '${key}' value ${value} is below minimum ${descriptor.min}.`,
        key,
        "VALIDATION_ERROR",
      );
    }
    if (descriptor.max !== undefined && value > descriptor.max) {
      throw new DotlyteError(
        `Environment variable '${key}' value ${value} exceeds maximum ${descriptor.max}.`,
        key,
        "VALIDATION_ERROR",
      );
    }
  }

  if (typeof value === "string") {
    if (descriptor.min !== undefined && value.length < descriptor.min) {
      throw new DotlyteError(
        `Environment variable '${key}' length ${value.length} is below minimum ${descriptor.min}.`,
        key,
        "VALIDATION_ERROR",
      );
    }
    if (descriptor.max !== undefined && value.length > descriptor.max) {
      throw new DotlyteError(
        `Environment variable '${key}' length ${value.length} exceeds maximum ${descriptor.max}.`,
        key,
        "VALIDATION_ERROR",
      );
    }
  }

  return value;
}

/** Check if a value looks like a Zod schema. */
function isZodSchema(value: unknown): boolean {
  return (
    value !== null &&
    typeof value === "object" &&
    "_def" in (value as Record<string, unknown>) &&
    "parse" in (value as Record<string, unknown>)
  );
}

/** Check if a value looks like a Valibot schema. */
function isValibotSchema(value: unknown): boolean {
  return (
    value !== null &&
    typeof value === "object" &&
    "type" in (value as Record<string, unknown>) &&
    "_run" in (value as Record<string, unknown>)
  );
}

/** Validate a value with a Zod schema. */
function validateWithZod(key: string, value: unknown, schema: { parse: (v: unknown) => unknown }): unknown {
  try {
    return schema.parse(value ?? undefined);
  } catch (err: unknown) {
    const message =
      err !== null && typeof err === "object" && "message" in err
        ? String((err as Record<string, unknown>).message)
        : String(err);
    throw new DotlyteError(
      `Validation failed for '${key}': ${message}`,
      key,
      "VALIDATION_ERROR",
    );
  }
}

/** Validate a value with a Valibot schema. */
function validateWithValibot(
  key: string,
  value: unknown,
  schema: { _run: (dataset: unknown, config: unknown) => unknown },
): unknown {
  try {
    // Valibot v1 uses safeParse or the _run internal method
    // Try to find safeParse at runtime for compatibility
    const result = schema._run(
      { typed: false, value: value ?? undefined },
      { abortEarly: true },
    ) as { typed: boolean; value: unknown; issues?: unknown[] };

    if (result.issues && result.issues.length > 0) {
      const issue = result.issues[0] as { message?: string };
      throw new DotlyteError(
        `Validation failed for '${key}': ${issue.message ?? "unknown error"}`,
        key,
        "VALIDATION_ERROR",
      );
    }

    return result.value;
  } catch (err) {
    if (err instanceof DotlyteError) throw err;
    throw new DotlyteError(
      `Validation failed for '${key}': ${err instanceof Error ? err.message : String(err)}`,
      key,
      "VALIDATION_ERROR",
    );
  }
}

/* ──────── Main API ──────── */

/**
 * Create a strongly-typed, validated configuration object.
 *
 * Validates at import-time (fail fast on app startup). Returns a frozen object
 * with full TypeScript type inference.
 *
 * Supports three schema formats:
 * - DotlyteSchema descriptors (built-in, zero dependencies)
 * - Zod schemas (auto-detected, optional peer dep)
 * - Valibot schemas (auto-detected, optional peer dep)
 *
 * Also supports server/client/shared sections for SSR framework safety.
 *
 * @example
 * ```ts
 * // Flat schema
 * const env = createTypedConfig({
 *   PORT: { type: 'integer', default: 3000 },
 *   DEBUG: { type: 'boolean', default: false },
 * })
 *
 * // Sectioned schema (Next.js / Nuxt)
 * const env = createTypedConfig({
 *   server: { DATABASE_URL: { type: 'string', required: true } },
 *   client: { NEXT_PUBLIC_APP_URL: { type: 'string', format: 'url' } },
 *   shared: { NODE_ENV: { type: 'string', enum: ['development', 'test', 'production'] } },
 * })
 * ```
 */
export function createTypedConfig<const S extends Record<string, unknown>>(
  schema: S,
  options: TypedConfigOptions = {},
): InferTypedConfig<S> {
  const isSectioned = "server" in schema && "client" in schema;

  if (isSectioned) {
    return createSectionedConfig(schema as unknown as SectionedSchema, options) as InferTypedConfig<S>;
  }

  return createFlatConfig(schema, options) as InferTypedConfig<S>;
}

/** Handle flat (non-sectioned) schemas. */
function createFlatConfig(
  schema: Record<string, unknown>,
  options: TypedConfigOptions,
): Record<string, unknown> {
  // Load raw values from all sources
  const config = load({
    ...options,
    defaults: options.defaults,
  });
  const raw = config.toObject();
  const envVars = process.env;

  const result: Record<string, unknown> = {};
  const errors: string[] = [];

  for (const [key, descriptor] of Object.entries(schema)) {
    try {
      // Read from: raw config (which includes env vars) > process.env directly
      const rawValue = raw[key.toLowerCase()] ?? raw[key] ?? envVars[key];

      if (isZodSchema(descriptor)) {
        result[key] = validateWithZod(key, rawValue, descriptor as { parse: (v: unknown) => unknown });
      } else if (isValibotSchema(descriptor)) {
        result[key] = validateWithValibot(
          key,
          rawValue,
          descriptor as { _run: (d: unknown, c: unknown) => unknown },
        );
      } else {
        result[key] = validateField(key, rawValue, descriptor as FieldDescriptor<TypeName, readonly unknown[], unknown, boolean>);
      }
    } catch (err) {
      if (options.skipValidation) {
        result[key] = raw[key.toLowerCase()] ?? raw[key] ?? envVars[key] ?? undefined;
      } else {
        errors.push(err instanceof Error ? err.message : String(err));
      }
    }
  }

  if (errors.length > 0) {
    throw new ValidationError(
      errors.map((msg, i) => ({
        key: Object.keys(schema)[i] ?? "unknown",
        rule: "typed-config",
        message: msg,
      })),
    );
  }

  // Freeze the result — make it truly immutable
  return Object.freeze(result);
}

/** Handle sectioned (server/client/shared) schemas. */
function createSectionedConfig(
  schema: SectionedSchema,
  options: TypedConfigOptions,
): Record<string, unknown> {
  const serverSchema = schema.server ?? {};
  const clientSchema = schema.client ?? {};
  const sharedSchema = schema.shared ?? {};
  const clientPrefix = schema.clientPrefix ?? "NEXT_PUBLIC_";

  // Validate that client keys start with the prefix
  for (const key of Object.keys(clientSchema)) {
    if (!key.startsWith(clientPrefix)) {
      throw new DotlyteError(
        `Client environment variable '${key}' must start with '${clientPrefix}'. ` +
          `Move it to the 'server' section or rename it.`,
        key,
        "VALIDATION_ERROR",
      );
    }
  }

  // Merge all into one flat schema for validation
  const allSchema = { ...serverSchema, ...clientSchema, ...sharedSchema };
  const result = createFlatConfig(allSchema, options);

  // Build sets for boundary enforcement
  const serverKeys = new Set(Object.keys(serverSchema));
  const clientKeys = new Set(Object.keys(clientSchema));
  const sharedKeys = new Set(Object.keys(sharedSchema));

  // Return a Proxy that enforces server/client boundaries
  return createBoundaryProxy(
    result as Record<string, unknown>,
    serverKeys,
    clientKeys,
    sharedKeys,
    options.onSecretAccess,
  );
}

/**
 * Extract the inferred config type from a schema.
 *
 * @example
 * ```ts
 * const schema = {
 *   PORT: { type: 'integer' as const, default: 3000 },
 *   DEBUG: { type: 'boolean' as const, default: false },
 * } as const
 *
 * type Env = InferConfig<typeof schema>
 * // { readonly PORT: number; readonly DEBUG: boolean }
 * ```
 */

File: langs/go/typed.go
package dotlyte

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

// FieldDescriptor defines the schema for a single typed configuration field.
type FieldDescriptor struct {
	// Type is the expected type: "string", "integer", "number", "boolean", "url".
	Type string

	// Required indicates the field must have a value. Defaults to true.
	Required bool

	// Default is the fallback value when the field is missing.
	Default interface{}

	// Enum restricts the value to one of these allowed values.
	Enum []interface{}

	// Min is the minimum value (for numbers) or minimum length (for strings).
	Min *float64

	// Max is the maximum value (for numbers) or maximum length (for strings).
	Max *float64

	// Sensitive marks the field as containing secret data (redacted in logs).
	Sensitive bool

	// Doc is a human-readable description of the field.
	Doc string
}

// TypedConfigOptions configures the behavior of CreateTypedConfig.
type TypedConfigOptions struct {
	// SkipValidation disables validation (useful in tests).
	SkipValidation bool

	// OnSecretAccess is called when a sensitive key is accessed (audit logging).
	OnSecretAccess func(key, context string)
}

// CreateTypedConfig creates a validated, typed configuration map from environment
// variables and a field schema. Values are coerced to their declared types and
// validated against constraints. Returns (nil, error) on validation failures.
func CreateTypedConfig(schema map[string]FieldDescriptor, opts ...TypedConfigOptions) (map[string]interface{}, error) {
	var options TypedConfigOptions
	if len(opts) > 0 {
		options = opts[0]
	}

	// Load config from environment
	config, err := Load(nil)
	if err != nil {
		return nil, err
	}
	raw := config.ToMap()
	envVars := os.Environ()
	envMap := make(map[string]string, len(envVars))
	for _, e := range envVars {
		if idx := strings.IndexByte(e, '='); idx >= 0 {
			envMap[e[:idx]] = e[idx+1:]
		}
	}

	result := make(map[string]interface{}, len(schema))
	var errors []string

	for key, desc := range schema {
		rawValue := lookupRawValue(raw, envMap, key)

		validated, err := validateTypedField(key, rawValue, desc)
		if err != nil {
			if options.SkipValidation {
				result[key] = rawValue
			} else {
				errors = append(errors, err.Error())
			}
			continue
		}
		result[key] = validated
	}

	if len(errors) > 0 {
		return nil, &ValidationError{
			DotlyteError: DotlyteError{
				Message: fmt.Sprintf("typed config validation failed with %d error(s): %s",
					len(errors), strings.Join(errors, "; ")),
				Code: "TYPED_CONFIG_VALIDATION_FAILED",
			},
		}
	}

	return result, nil
}

// CreateSectionedConfig creates a typed configuration with server/client/shared
// sections for SSR framework boundary enforcement. Server keys are restricted,
// client keys must start with clientPrefix, and shared keys are available everywhere.
func CreateSectionedConfig(
	server, client, shared map[string]FieldDescriptor,
	clientPrefix string,
	opts ...TypedConfigOptions,
) (map[string]interface{}, error) {
	if clientPrefix == "" {
		clientPrefix = "NEXT_PUBLIC_"
	}

	// Validate client keys start with prefix
	for key := range client {
		if !strings.HasPrefix(key, clientPrefix) {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("client environment variable '%s' must start with '%s'", key, clientPrefix),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
	}

	// Merge all schemas and validate
	allSchema := make(map[string]FieldDescriptor)
	for k, v := range server {
		allSchema[k] = v
	}
	for k, v := range client {
		allSchema[k] = v
	}
	for k, v := range shared {
		allSchema[k] = v
	}

	result, err := CreateTypedConfig(allSchema, opts...)
	if err != nil {
		return nil, err
	}

	// Build boundary key sets
	serverKeys := make(map[string]bool, len(server))
	for k := range server {
		serverKeys[k] = true
	}
	clientKeys := make(map[string]bool, len(client))
	for k := range client {
		clientKeys[k] = true
	}
	sharedKeys := make(map[string]bool, len(shared))
	for k := range shared {
		sharedKeys[k] = true
	}

	var onSecretAccess func(string, string)
	if len(opts) > 0 {
		onSecretAccess = opts[0].OnSecretAccess
	}

	// Wrap in boundary config and return the underlying data
	bc := NewBoundaryConfig(result, serverKeys, clientKeys, sharedKeys, onSecretAccess)
	return bc.data, nil
}

// lookupRawValue tries to find a value by key in the loaded config and env vars.
func lookupRawValue(raw map[string]interface{}, envMap map[string]string, key string) interface{} {
	// Try loaded config first (case-insensitive)
	if v, ok := raw[strings.ToLower(key)]; ok {
		return v
	}
	if v, ok := raw[key]; ok {
		return v
	}
	// Try environment variables directly
	if v, ok := envMap[key]; ok {
		return v
	}
	return nil
}

// validateTypedField validates and coerces a single field value against its descriptor.
func validateTypedField(key string, raw interface{}, desc FieldDescriptor) (interface{}, error) {
	isRequired := desc.Required
	hasDefault := desc.Default != nil

	// Apply default
	value := raw
	if isEmpty(value) && hasDefault {
		value = desc.Default
	}

	// Check required
	if isEmpty(value) && isRequired {
		msg := fmt.Sprintf("missing required environment variable '%s'", key)
		if desc.Doc != "" {
			msg += fmt.Sprintf(" (%s)", desc.Doc)
		}
		msg += ". Set it in your .env file, config file, or as an environment variable."
		return nil, &MissingRequiredKeyError{
			DotlyteError: DotlyteError{
				Message: msg,
				Key:     key,
				Code:    "MISSING_REQUIRED_KEY",
			},
		}
	}

	// Optional and absent
	if isEmpty(value) {
		return nil, nil
	}

	// Coerce to target type
	coerced, err := coerceToType(key, value, desc.Type)
	if err != nil {
		return nil, err
	}

	// Enum validation
	if len(desc.Enum) > 0 {
		found := false
		for _, allowed := range desc.Enum {
			if fmt.Sprintf("%v", coerced) == fmt.Sprintf("%v", allowed) {
				found = true
				break
			}
		}
		if !found {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' must be one of %v, got '%v'", key, desc.Enum, coerced),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
	}

	// Min/Max for numbers
	if num := toFloat64(coerced); num != nil {
		if desc.Min != nil && *num < *desc.Min {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' value %v is below minimum %v", key, coerced, *desc.Min),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
		if desc.Max != nil && *num > *desc.Max {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' value %v exceeds maximum %v", key, coerced, *desc.Max),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
	}

	// Min/Max for strings (length)
	if s, ok := coerced.(string); ok {
		if desc.Min != nil && float64(len(s)) < *desc.Min {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' length %d is below minimum %v", key, len(s), *desc.Min),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
		if desc.Max != nil && float64(len(s)) > *desc.Max {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' length %d exceeds maximum %v", key, len(s), *desc.Max),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
	}

	return coerced, nil
}

// coerceToType converts a value to the specified type.
func coerceToType(key string, value interface{}, typeName string) (interface{}, error) {
	strValue := fmt.Sprintf("%v", value)

	switch typeName {
	case "boolean":
		if b, ok := value.(bool); ok {
			return b, nil
		}
		lower := strings.ToLower(strValue)
		switch lower {
		case "true", "yes", "1", "on":
			return true, nil
		case "false", "no", "0", "off":
			return false, nil
		}
		return nil, &DotlyteError{
			Message: fmt.Sprintf("environment variable '%s' expected boolean, got '%s'", key, strValue),
			Key:     key,
			Code:    "VALIDATION_ERROR",
		}

	case "integer":
		if i, ok := value.(int64); ok {
			return i, nil
		}
		if i, ok := value.(int); ok {
			return int64(i), nil
		}
		if f, ok := value.(float64); ok {
			if f == float64(int64(f)) {
				return int64(f), nil
			}
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' expected integer, got '%s'", key, strValue),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
		i, err := strconv.ParseInt(strValue, 10, 64)
		if err != nil {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' expected integer, got '%s'", key, strValue),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
		return i, nil

	case "number":
		if f, ok := value.(float64); ok {
			return f, nil
		}
		if i, ok := value.(int64); ok {
			return float64(i), nil
		}
		if i, ok := value.(int); ok {
			return float64(i), nil
		}
		f, err := strconv.ParseFloat(strValue, 64)
		if err != nil {
			return nil, &DotlyteError{
				Message: fmt.Sprintf("environment variable '%s' expected number, got '%s'", key, strValue),
				Key:     key,
				Code:    "VALIDATION_ERROR",
			}
		}
		return f, nil

	case "string", "url":
		s := strValue
		if typeName == "url" {
			if !strings.Contains(s, "://") {
				return nil, &DotlyteError{
					Message: fmt.Sprintf("environment variable '%s' is not a valid URL: '%s'", key, s),
					Key:     key,
					Code:    "VALIDATION_ERROR",
				}
			}
		}
		return s, nil

	default:
		// Unknown type, return as-is
		return value, nil
	}
}

// isEmpty checks if a value is nil, empty string, or absent.
func isEmpty(v interface{}) bool {
	if v == nil {
		return true
	}
	if s, ok := v.(string); ok {
		return s == ""
	}
	return false
}
